pkg/volume/flexvolume/flexvolume.go

```diff
@@ -91,61 +92,71 @@ func (plugin *flexVolumePlugin) GetAccessModes() []api.PersistentVolumeAccessMod
 	}
 }
 
-func (plugin *flexVolumePlugin) getVolumeSource(spec *volume.Spec) *api.FlexVolumeSource {
-	var source *api.FlexVolumeSource
-	if spec.Volume != nil && spec.Volume.FlexVolume != nil {
-		source = spec.Volume.FlexVolume
-	} else if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.FlexVolume != nil {
-		source = spec.PersistentVolume.Spec.FlexVolume
+// Attach interface
+
+func (plugin *flexVolumePlugin) Attach(spec *volume.Spec, hostName string) error {
+	call := NewDriverCall(plugin.execPath, attachCmd, exec.New())
+	call.AppendSpec(spec)
+	call.Append(hostName)
+
+	_, err := call.Run()
+	return err
+}
+
+func (plugin *flexVolumePlugin) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+	call := NewDriverCallWithTimeout(plugin.execPath, waitForAttachCmd, exec.New(), timeout)
```

Comment:
We do not need "waitForAttachCmd" call. I like your initial implementation, where attach is non-blocking and waitForAttach waits on a channel. We can kill the process on timeout to avoid any leaks.